### PR TITLE
Add OS native interface lib, with objective-c++ interface to Mac Cocoa ibraries

### PR DESCRIPTION
**Pull request by dakcarto**
This library is intended to offer abstraction to the host OS's underlying public interfaces. This is useful for OSes that provide interfaces in languages other than C/C++, or for grouping calls to OS-specific code so that it only needs to be updated in one place in the source tree and handles any deprecation gracefully. It is advisable to leverage existing functions provided by Qt, rather than rely upon OS-specific code, unless such code extends the application to provide a better OS-specific user experience or solve a problem. This lib provides such a solution.

**Example first use case**

As of Mac OS X 10.9 (Mavericks) many system public API calls to Carbon libraries (based upon C) have been deprecated in favor of modern Cocoa libraries (written in Objective-C), which can no longer be directly called from C++. Coalescing and mixing these new calls in a library, using Objective-C++ allows not only access to the Apple system Objective-C libraries [and frameworks](https://developer.apple.com/library/mac/navigation/#section=Frameworks), but also those from third-parties, like the auto-updating [Sparkle.framework](http://sparkle.andymatuschak.org/).

See also: http://el-tramo.be/blog/mixing-cocoa-and-qt/

Once available, probably all Mac OS X specific code can be moved out of the QGIS source tree and to the new library, and switched to generic interface calls.

Includes initial unit test for Mac native lib.

### CODE REVIEW

**Comment by nyalldawson** [comment]
@dakcarto what's the status of this PR? I gather something like this is still required in the long run, but this particular patch is quite old now. Is this safe to close?

**Comment by dakcarto** [comment]
Hi,

So, this is rather important for current and future Mac support, since Carbon support (C API) is deprecated in favor of Cocoa (Objective-C API). It can also help fix issues like [redmine # 11399](http://hub.qgis.org/issues/11399), while allowing simple support for both Qt4 -> 5 and Mac OS upgrade changes in one place.

Should I rework this for 2.12 to fix Mac bugs and deprecation compiling notices or just wait until after the feature freeze?

**Comment by jef-n** [comment]
@dakcarto I wonder why this wasn't committed long ago -  in 2.5, 2.7, 2.9 or 2.11.  It looks safe and doesn't touch other platforms AFAICT.  So if it fixes mac bugs, it's your call.  You're our OSX expert anyway.   All the talk just makes me wonder more, where the catch is ;)

**Comment by dakcarto** [comment]
@jef-n wrote:
> @dakcarto I wonder why this wasn't committed long ago - in 2.5, 2.7, 2.9 or 2.11. It looks safe and doesn't touch other platforms AFAICT. So if it fixes mac bugs, it's your call. You're our OSX expert anyway. All the talk just makes me wonder more, where the catch is ;)

No catch (that I know of)  :^). I got busy with other projects (and two toddlers!) for a year or so and wasn't able to commit much time to QGIS, let alone commit actual code. Basically, I kind of forgot about this PR.

I can rework it, and if it is functional, are you OK with me merging it for 2.12? In its current state it adds no features, just introduces Objective-C++ to the mix. Later on, it can be leveraged to add OS-specific features that Qt does not provide.

**Comment by jef-n** [comment]
@dakcarto wrote:
> I can rework it, and if it is functional, are you OK with me merging it for 2.12?

That's what "it's your call" was supposed to say (doesn't it?  I'm not a native speaker). Just put the emphasis on the it-fixes-bugs-on-osx part ;)

**Comment by dakcarto** [comment]
Okey-dokey. Thanks for the clarification!

**Comment by michaelkirk** [comment]
:+1: 

This branch has conflicts now, but @dakcarto do reach out when you're ready if you want a sanity check, I can also build this.

File: src/app/qgisapp.cpp [hunk 1]
```diff
@@ -83,6 +83,12 @@
 #ifdef Q_OS_MACX
 #include <ApplicationServices/ApplicationServices.h>
 
+// Virtual interfaces to Cocoa objective-c frameworks/classes/calls
+// cocoainitializer is to handle objective-c garbage collection
+//   see: http://el-tramo.be/blog/mixing-cocoa-and-qt/
+//#include "cocoainitializer.h"
+#include "qgsmacappkit.h"
```
Comment (m-kuhn) [inline src/app/qgisapp.cpp]: Shouldn't this live in an `#ifdef`?

File: src/native/README [hunk 2]
```diff
@@ -0,0 +1,23 @@
+README for qgis_native lib
```
Comment (m-kuhn) [inline src/native/README]: better to rename this file to `README.md` for pretty printing on github

File: src/native/mac/qgsmacnative.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSMACNATIVE_H
+#define QGSMACNATIVE_H
+
+class QgsMacAppKit
```
Comment (m-kuhn) [inline src/native/mac/qgsmacnative.h]: Hmmm... I was expecting this in qgsmacappkit.h.

File: src/core/CMakeLists.txt [hunk 4]
```diff
@@ -579,6 +579,12 @@ INCLUDE_DIRECTORIES(
   ${EXPAT_INCLUDE_DIR}
 )
 
+IF(USE_NATIVE_LIB)
+  IF(APPLE)
```
Comment (m-kuhn) [inline src/core/CMakeLists.txt]: Just wondering... Does the core library need to know about mac specific includes?
I assume if there's a base class interface defined with empty method bodies which can be overwritten in platform specific subclasses, all this logic could be kept separate in the src/native folder

**Comment by nyalldawson** [comment]
@dakcarto @m-kuhn can this be closed in favour of #3964 ? This PR queue is getting ridiculous :)

**Pull request closed without merging**